docs/self-managed/zeebe-deployment/security/client-authorization.md

```diff
@@ -226,12 +159,100 @@ func main() {
 Like the Java client, the Go client will not prevent you from adding credentials to gRPC calls while using an insecure connection, but doing so will expose your access token.
 :::
 
-### Environment variables
+#### Environment variables
 
 Since there are several environment variables that can be used to configure an `OAuthCredentialsProvider`, we list them here along with their uses:
 
 - `ZEEBE_CLIENT_ID` - The client ID used to request an access token from the authorization server
 - `ZEEBE_CLIENT_SECRET` - The client secret used to request an access token from the authorization server
-- `ZEEBE_TOKEN_AUDIENCE` - The address for which the token should be valid
+- `ZEEBE_TOKEN_AUDIENCE` - The audience for which the token should be valid
 - `ZEEBE_AUTHORIZATION_SERVER_URL` - The URL of the authorization server from which the access token will be requested (by default, configured for Camunda Platform 8)
 - `ZEEBE_CLIENT_CONFIG_PATH` - The path to a cache file where the access tokens will be stored (by default, it's `$HOME/.camunda/credentials`)
+
+### Custom Credentials provider
+
+As previously mentioned, the `CredentialProvider`'s purpose is to modify the gRPC headers with an authorization method such that a reverse proxy sitting in front of the gateway can validate them.
+
+The interface consists of an `applyCredentials` method and a `shouldRetryRequest` method. The first method is called for each gRPC call and takes a map of headers to which it should add credentials. The second method is called whenever a gRPC call fails and takes in the error that caused the failure which is then used to decide whether the request should be retried.
+
+The following sections implement simple custom provider in Java and Go.
+
+#### Java
+
+```java
+public class MyCredentialsProvider implements CredentialsProvider {
+    /**
+     * Adds a token to the Authorization header of a gRPC call.
+    */
+    @Override
+    public void applyCredentials(final Metadata headers) {
+      final Key<String> authHeaderkey = Key.of("Authorization", Metadata.ASCII_STRING_MARSHALLER);
+      headers.put(authHeaderKey, "Bearer someToken");
+    }
+
+    /**
+    * Retries request if it failed with a timeout.
+    */
+    @Override
+    public boolean shouldRetryRequest(final Throwable throwable) {
+      return ((StatusRuntimeException) throwable).getStatus() == Status.DEADLINE_EXCEEDED;
+    }
+}
+```
+
+After implementing the `CredentialsProvider`, we can provide it when building a client:
+
+```java
+public class SecureClient {
+    public static void main(final String[] args) {
+      final ZeebeClient client = ZeebeClient.newClientBuilder().credentialsProvider(new MyCredentialsProvider()).build();
+
+      // continue...
+    }
+}
+```
+
+#### Go
+
+```go
+package main
+
+import (
+    "context"
+    "fmt"
+    "google.golang.org/grpc/status"
+    "google.golang.org/grpc/codes"
+    "github.com/camunda-cloud/zeebe/clients/go/pkg/zbc"
```

Comment:
Is this the correct import statement? Or should it be `/camunda/`?

Reply:
that appears outdated indeed, so we may also fix that in the older versions (I just moved this section without modifying it)

Reply:
I'll create an issue and we can handle that separately. Thanks!